src/librustc_resolve/resolve_imports.rs

```diff
@@ -708,10 +708,18 @@ impl<'a, 'b:'a, 'tcx:'b> ImportResolver<'a, 'b, 'tcx> {
                                                       import_directive.span,
                                                       name,
                                                       ns);
-                    dest_import_resolution.id = id;
-                    dest_import_resolution.is_public = is_public;
-                    dest_import_resolution.target = Some(target.clone());
-                    self.add_export(module_, name, &dest_import_resolution);
+                    let other_ns = if ns == TypeNS { ValueNS } else { TypeNS };
+                    let shadowed = module_.get_child(name, ns).is_some() || {
+                        module_.get_child(name, other_ns).is_some() &&
+                        target.binding.defined_with(DefModifiers::LINKED_NAMESPACES)
+                    };
+
+                    if !shadowed {
+                        dest_import_resolution.id = id;
+                        dest_import_resolution.is_public = is_public;
+                        dest_import_resolution.target = Some(target.clone());
+                        self.add_export(module_, name, &dest_import_resolution);
+                    }
```

Comment:
So, reexport sets are free from name conflicts now and https://github.com/rust-lang/rust/pull/31308 can be landed?